Core/clim-basic/extended-streams/stream-input.lisp

```diff
@@ -179,58 +179,72 @@
+;;;
+;;; Part of the extended input stream protocol is mixed in thanks to
+;;; the input-stream-kernel (STREAM-INPUT-BUFFER, STREAM-SCAN-POINTER
+;;; and STREAM-INSERTION-POINTER accessors and :INPUT-BUFFER
+;;; initarg). It is good because these parts makes sense here.
 
-(defclass standard-input-stream (input-kernel-mixin
+(defclass standard-input-stream (input-stream-kernel
                                  input-stream
-                                 standard-sheet-input-mixin)
-  ((unread-chars :initform nil :accessor stream-unread-chars)))
-
-;;; XXX: fixing stream shisophrenia is a subject of the next input-refactor pass.
-;;; 1. What about EOF?
-;;; 2. GESTURE-OBJECT should be already coerced to a character!
-;;;   2a. We don't handle modifiers here, so it is double wrong, see stream-process-gesture.
-;;; 3. HANDLE-EVENT should have been invoked on a completely different level. -- jd 2018-12-20
-(defmethod stream-read-char ((pane standard-input-stream))
-  (if (stream-unread-chars pane)
-      (pop (stream-unread-chars pane))
-      (let ((event (event-read pane)))  ; (1)
-        (if (and (typep event 'key-press-event)  ; (2)
-                 (keyboard-event-character event)) ; (2a)
-            (let ((char (char-for-read (keyboard-event-character event))))
-              (stream-write-char pane char)
-              (return-from stream-read-char char))
-            ;; (3)
-            (handle-event (event-sheet event) event)))))
-
-(defmethod stream-unread-char ((pane standard-input-stream) char)
-  (push char (stream-unread-chars pane)))
-
-(defmethod stream-read-char-no-hang ((pane standard-input-stream))
-  (if (stream-unread-chars pane)
-      (pop (stream-unread-chars pane))
-      (loop for event = (event-read-no-hang pane)
-            if (null event)
-              return nil
-            if (and (typep event 'key-press-event)
-                    (keyboard-event-character event))
-              return (char-for-read (keyboard-event-character event))
-            else
-              do (handle-event (event-sheet event) event))))
-
-(defmethod stream-clear-input ((pane standard-input-stream))
-  (setf (stream-unread-chars pane) nil)
-  (loop for event = (event-read-no-hang pane)
-        if (null event)
-          return nil
-        else
-          do (handle-event (event-sheet event) event))
-  nil)
+                                 fundamental-character-input-stream)
+  ())
+
+(defmethod stream-append-gesture :before ((stream standard-input-stream) gesture)
+  (check-type gesture character))
+
+(defmethod handle-event :after
+    ((client standard-input-stream)
+     (event key-press-event))
+  (when-let ((ch (event-char event)))
+    (stream-append-gesture client ch)))
+
+(defmethod stream-process-gesture
+    ((stream standard-input-stream) gesture type)
+  (declare (ignore type))
+  (if-let ((char (and (typep gesture 'key-press-event)
+                      (event-char gesture))))
+    (values char 'character)
+    nil))
```

Comment:
mind, that input buffers may be shared among multiple streams. I can imagine that one input buffer is shared by both SIS and SEIS, then when we read a gesture from SIS, then everything but characters should be ignored.

Regarding tests: that would require extending the Null port or faking the event queue for the stream, because that were events come in (via handle-event poked from stream-input-wait).

Reply:
Injecting a mock event queue seems like the better option.

Reply:
I've added a smoke test without testing interactions with the event queue (so only some brief input buffer tests).

Reply:
Great.